### PR TITLE
Generate the stats.End event when a client stream finishes

### PR DESCRIPTION
**Pull request by charithe**
When used with a client-only streaming call, `stats.Handler` does not receive the `End` event if the call ends successfully. This seems to be due to the fact that there's no opportunity on the server side to produce an EOF after the client calls the generated `CloseAndRecv()` method on the stream.

This patch adds an extra check to the `RecvMsg` function on the client stream implementation to detect whether the stream is closed and trigger the `End` event generation at the end of the function. As the current tests only cover unary calls and bidi streams, tests for server streams and client streams have been added as well.

### CODE REVIEW

**Comment by charithe** [comment]
I signed it.

**Comment by menghanl** [comment]
Thanks for your fix.
The problem exists not only for client streaming, but for all streaming RPCs.
This should been fixed by moving stats.End to clientStream.finish() (#1182)

The tests you added would be a good test for situation like this.
If you don't mind, can you resolve the conflicts in the test so we can get the tests merged? Thanks.

**Comment by charithe** [comment]
Great to hear that the issue has been fixed. I resolved the merge conflicts but, oddly, the Travis build fails on lint checks. They all succeed on my machine so I am not quite sure what the problem is. I'll investigate further tomorrow.

File: stream.go [hunk 1]
```diff
@@ -172,6 +172,16 @@ func newClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 			}
 		}()
 	}
+	defer func() {
```
Comment (menghanl) [inline stream.go]: Can you also remove this? I believe this is a git merge error.

**Comment by charithe** [comment]
Removed the merge artifacts and also added back a test that I had missed during the merge.

**Pull request merged**